### ISSUE TITLE
Tracking issue for RFC 2351, "Add `is_sorted` to the standard library"

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#53485]
Feature gate: `#![feature(is_sorted)]`

This is a tracking issue for `is_sorted{_by,_by_key}` functions on `[T]` and `Iterator` (rust-lang/rfcs#2351).

### Public API

```rust
impl<T> [T] {
    pub fn is_sorted(&self) -> bool
    where
        T: PartialOrd;

    pub fn is_sorted_by<'a, F>(&'a self, mut compare: F) -> bool
    where
        F: FnMut(&'a T, &'a T) -> bool;

    pub fn is_sorted_by_key<'a, F, K>(&'a self, f: F) -> bool
    where
        F: FnMut(&'a T) -> K,
        K: PartialOrd;
}
```
```rust
// core::iter

pub trait Iterator {
    // all the other methods omitted

    fn is_sorted(self) -> bool
    where
        Self: Sized,
        Self::Item: PartialOrd;

    fn is_sorted_by<F>(mut self, compare: F) -> bool
    where
        Self: Sized,
        F: FnMut(&Self::Item, &Self::Item) -> bool;

    fn is_sorted_by_key<F, K>(self, f: F) -> bool
    where
        Self: Sized,
        F: FnMut(Self::Item) -> K,
        K: PartialOrd;
}
```

### Steps / History

- [x] Implementation: #55045
- [x] #62473
- [x] #102977
- [x] #112699
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- [x] [Require `Ord` instead of only `PartialOrd`?](https://github.com/rust-lang/rfcs/blob/master/text/2351-is-sorted.md#require-ord-instead-of-only-partialord)
    - Answered by https://github.com/rust-lang/rust/pull/81382#issuecomment-776672739 (no, `PartialOrd` is the right bound)
- [x] [Should `Iterator::is_sorted_by_key` be added as well?](https://github.com/rust-lang/rfcs/blob/master/text/2351-is-sorted.md#should-iteratoris_sorted_by_key-be-added-as-well)
    - It was added in the original implementation PR
- [ ] [Add `std::cmp::is_sorted` instead?](https://github.com/rust-lang/rfcs/blob/master/text/2351-is-sorted.md#add-stdcmpis_sorted-instead)
- [x] Should `is_sorted_by` take a closure returning `bool` instead of `Option<Ordering>`?
  - Added in #118811

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by LukasKalbertodt** [comment]
@gnzlbg Would you be interested in implementing this? You basically already [wrote the code](https://github.com/gnzlbg/is_sorted). I think it would be fancy to have this SIMD code in the standard library. (Or should we start with only the fully generic algorithm?). PS: Sorry for not having reviewed https://github.com/rust-lang/rust/pull/53386 yet -- I hope I find the time and mental capacity to do that in the near future!

@Centril The links for the unresolved questions are dead (still pointing to the 0000 file). Could you or anyone update them? :)

---

Regarding the "**Should `Iterator::is_sorted_by_key` be added as well?**" question: [this last minute comment in the RFC thread](https://github.com/rust-lang/rfcs/pull/2351#issuecomment-414039302) argues that this method should be added:

> I feel like it is not a question about how easy it is to do it yourself using map but more about completeness of the api.

And people seem to agree. So I'd also say that this method should be added.

**Comment by gnzlbg** [comment]
> You basically already wrote the code.

I have commented in the issue.

**Comment by SimonSapin** [comment]
Oh, I didn’t realized that `Iterator::is_sorted_by_key` was not part of the RFC when I proposed FCP. I thought there was consensus to include it for consistence, both internal and with the existing `sort_by_key`/`min_by_key`/`max_by_key` methods which are also redundant with their respective method like `sort_by`.

We do have precedent of adding *some* APIs that can be expressed as a terse combination of existing APIs: https://github.com/rust-lang/rust/pull/49098

@clarcharr, you expressed the opposite in https://github.com/rust-lang/rfcs/pull/2351#issuecomment-372083545. Do you still think it should not be included?

**Comment by clarcharr** [comment]
At the time I made that comment, I didn't realise that `max_by_key` and `min_by_key` were already stable (as of 1.6, which is a while ago). Basically, there are two options:

1. Deprecate `max_by_key` and `min_by_key` and don't add `is_sorted_by_key`.
2. Add `is_sorted_by_key`.

I feel like the usefulness of these should be re-evaluated, as it forces the key to outlive any reference to `self` rather than just the current reference to `self`. Basically, we'd need something similar to GAT for it to be useful, and `map` already exists.

Personally, I'd prefer option 1. But option 2 is also acceptable and there may be a way to make these functions work with GAT in the future.

**Comment by gnzlbg** [comment]
@SimonSapin would you like to be the reviewer of a PR implementing this ? If so, I'd like to discuss with somebody first how the is_sorted crate is structure, and how to best implement this functionality in the std library.

**Comment by gnzlbg** [comment]
So I don't know how to properly implement this in `core`. 

The `is_sorted` functionality should use intrinsics when available at run-time. The `iter` and `slice` modules in `core` do not have access to run-time detection (that requires `std`), but e.g. the `Iterator` trait exported by `core` and `std` has to be the same, so we can't override the `core` implementation in `std` to add run-time detection on top. 

After talking with @SimonSapin on IRC, I am just going to add the naive versions to `core` which get picked up by `std` and call it a day. Those who need more performance can just use `is_sorted` from crates.io. Adding vectorized versions of the `is_sorted` functionality isn't worth it either, because one would need to re-compile `core` to profit from it (using crates.io is just easier). Maybe if we get newer targets in the future with more features enabled this might become worth it.

**Comment by kleimkuhler** [comment]
I have interest in working on this RFC. This would be a significant learning experience for me, but I don't see any reason why I should not be able to complete it.

@gnzlbg If your offer for being available as a mentor is still open, I'd like to take you up on that!

If there is any concern in timeline (I should be able to begin implementing this week) or preference for someone who may have more experience, I have no problem with that; just let me know!

**Comment by gnzlbg** [comment]
My recommendation is that if you want to implement this now, you should just adapt https://github.com/bluss/rust-itertools/pull/261 .

An efficient implementation of this is probably blocked on https://github.com/rust-lang/rfcs/pull/2492

**Comment by kleimkuhler** [comment]
Thanks for the links; it makes sense that a more efficient implementation is blocked on that right now.

To clarify on my initial comment (and what you picked up on in your last sentence), is the "learning" part of this would be more related to familiarizing myself with the API and codebase.

I think an initial implementation based off https://github.com/bluss/rust-itertools/pull/261 could be a good first step and we'll keep an eye on the progress of https://github.com/rust-lang/rfcs/pull/2492?

**Comment by kleimkuhler** [comment]
@gnzlbg You'll find my initial implementation in the PR above (https://github.com/rust-lang/rust/pull/55045). I'm not sure who should be officially reviewing this, but I can update the PR if need be!

**Comment by kleimkuhler** [comment]
@SimonSapin It was asked a few comments up if you'd like to be the reviewer of this PR? I still need to assign someone to review it, but I'm just not sure who.

**Comment by clarcharr** [comment]
Call IsSorted::is_sorted directly.

**Comment by mtak-** [comment]
@clarcharr The above error happens when adding `is_sorted` as a dependency (without using it). The `is_sorted` crate itself no longer builds.

**Comment by CryZe** [comment]
You can send a PR to the is_sorted crate and use the patch section in your Cargo.toml in the meantime until it is merged and published.

**Comment by mtak-** [comment]
@gnzlbg No worries. Thanks for responding so quickly!

**Comment by timvermeulen** [comment]
As I accidentally commented on the implementation PR thread instead of here:
- Would it be better to implement `Iterator::is_sorted_by_key` in terms of `map` and `is_sorted` rather than delegating it to `is_sorted_by`? It seems to me that `f` is currently called more often than necessary.
- Would it make sense to give `Iterator::is_sorted_by_key`'s `F` the bound `FnMut(Self::Item) -> K` instead of `FnMut(&Self::Item) -> K`? Unlike `max_by_key`/`min_by_key`, `is_sorted_by_key` doesn't need to retain any of the iterator's values, so we probably don't need to pass them by reference.

**Comment by timvermeulen** [comment]
Something else to ponder: the closure parameter of `is_sorted_by` has the return type `Option<Ordering>`, which is a little awkward – it seems only useful for when you want to call `PartialOrd::partial_cmp` from it, but it's doubtful how common that will be given that we also have `is_sorted` and `is_sorted_by_key`.

Let's look at how the return value of the closure affects the outcome of `is_sorted_by`:
- If it returns `Some(Ordering::Less)` or `Some(Ordering::Equal)`, the iterator is still sorted up to that point and iteration continues.
- If it returns `Some(Ordering::Greater)` or `None`, the iterator is not sorted and the method returns `false`.

So either the two given items are in sorted order, or they're not. Shouldn't it just return a `bool`? It seems like it would make some common uses much simpler: checking that an iterator is in, say, strictly descending order would just be a matter of calling `.is_sorted_by(PartialOrd::gt)`. In general, any previous instances of `a.cmp(b)` return values in that closure could be replaced with `a <= b`.

The only downside I see is that it might make this method _too_ useful – it could be used to check conditions that don't have anything to do with sorting like checking that all items are equal, or that no two consecutive items are equal, or that the difference between any two consecutive items doesn't exceed some limit, etc.

**Comment by cuviper** [comment]
> The only downside I see is that it might make this method _too_ useful – it could be used to check conditions that don't have anything to do with sorting like [...]

You can already kludge things like that with `Ordering` if you really want. I think that's mostly OK, but perhaps there would be room for a more generic name like `compare_by`, or maybe `all`/`any` analogues on `&Item` pairs. See also `is_partitioned()` that I've proposed in #62278.

**Comment by timvermeulen** [comment]
>You can already kludge things like that with Ordering if you really want.

Good point!

>perhaps there would be room for a more generic name

I think the more generic way to do this would be to create an iterator over pairs of consecutive elements and then call `all` on that. We would still probably want an `is_sorted_by` method, though, and I think it would still benefit from having its closure parameter return a `bool`.

**Comment by KodrAus** [comment]
> Require Ord instead of only PartialOrd?

> the closure parameter of is_sorted_by has the return type Option<Ordering>, which is a little awkward

Personally, I was expecting an `Ord` bound on these methods rather than `PartialOrd` and for `is_sorted_by` to return `Ordering` rather than `Option<Ordering>`. That doesn't help folks wanting to sort floats much, but leaves them in at least a consistently incompatible place with respect to the existing sorting methods.

**Comment by petrochenkov** [comment]
I recently needed to check a slice for sortedness and found this function pretty useless because saying "not sorted" is usually not enough, you need to return a position at which the collection is not sorted to do anything useful.

**Comment by petrochenkov** [comment]
In other words, if this function needs to exist in the standard library, it should probably exist as a convenience on top of some other function, like `contains` is a convenience over `get`.

**Comment by KodrAus** [comment]
@petrochenkov I guess to align with `.binary_search()` in that case those underlying methods could return a `Result<(), usize>`, with `is_sorted` effectively becoming a call to `.sorting().ok()`?

**Comment by KodrAus** [comment]
I’m wondering whether there’s a return for the sorting methods returning an index that’s intuitive enough to make the ones returning a `bool` unnecessary.

**Comment by KodrAus** [comment]
With #72568 I think an `Ord` bound is quite compelling.

**Comment by LukasKalbertodt** [comment]
> With #72568 I think an `Ord` bound is quite compelling. 

I agree, we should just stabilize it with `Ord` (as soon as all other open questions are resolved). Then we can also change the `Option<Ordering>` to just `Ordering`

> I recently needed to check a slice for sortedness and found this function pretty useless because saying "not sorted" is usually not enough, you need to return a position at which the collection is not sorted to do anything useful.

Can you share more details about your use case? The RFC lists unit tests and "asserts" as the main reason why `is_sorted` would be useful. If you have another use case, that would be interesting to know. However, I would tend to keep these methods simple instead of making them general enough for every possible use case.

**Comment by TheWastl** [comment]
Regarding the `is_sorted_by_key` question: It seems that `is_sorted_by_key` would be preferred to `is_partitioned` - see #80085.

**Comment by LunarLambda** [comment]
I would like to have `is_sorted_by_key`, for sure.

I'm using `binary_search_by_key` to make sure I always insert elements into a Vec in a way that preserves a sorted order, but just to be absolutely sure I wanted to throw in a `debug_assert!(list.is_sorted_by_key(...))`, which led me here

**Comment by LukasKalbertodt** [comment]
> I'm using `binary_search_by_key` to make sure I always insert elements into a Vec in a way that preserves a sorted order, but just to be absolutely sure I wanted to throw in a `debug_assert!(list.is_sorted_by_key(...))`, which led me here

Whether or not to add/stabilize `is_sorted_by_key` is only about the `Iterator` method, not the slice method. Since you mentioned `Vec`, I assume you referred to `[T]::is_sorted_by_key` and your point does not apply anymore to the `Iterator::is_sorted_by_key` discussion. Is that correct? If I'm wrong, could you explain why you would prefer `iterator.is_sorted_by_key(...)` over `iterator.map(...).is_sorted()`?

**Comment by aobatact** [comment]
Why not return `Option<bool>` with `PartialOrd`?
If some `partial_cmp` return `None` during the check, we can return `None` instead.

**Comment by clarfonthey** [comment]
I can't imagine why it would be super useful to return `None` instead of `false`. If, for example, you have a `NaN` in your list, it's not sorted; saying it's "unsortable" would not really provide any extra information.

**Comment by aobatact** [comment]
Well knowing why it is not sorted is useful (at least for me).
For example we can do like this.
```
match slice.is_sorted(){
    Some(true) => {}//nop
    Some(false) => {
        slice.sort();
    }
    None => {/*remove Nan or return Error */},
}
// use the sorted slice
```
But even if is_sorted returns Some(false), NaN could still be included after it, so you could argue that it's meaningless.

Well I came up with other idea like having `is_sorted_until(&self) -> Result<(),Result<usize,usize>>`, though it will be other issue.

**Comment by kriomant** [comment]
I tried to use `is_sorted_by_key` [this way](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=93317ebee5d5f4cc78628e90300879e0):

```rust
#![feature(is_sorted)]

struct A {
    name: String
}

fn main() {
    let a = &[A{name: "1".to_string()}, A{name: "2".to_string()}];
    assert!(a.iter().is_sorted_by_key(|a| a.name.as_str()));
}
```

and got error:

```
error: lifetime may not live long enough
 --> src/lib.rs:9:36
  |
9 |     assert!(a.is_sorted_by_key(|a| a.name.as_str()));
  |                                 -- ^^^^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
  |                                 ||
  |                                 |return type of closure is &'2 str
  |                                 has type `&'1 A`
```

Is it possible to make it work?

**Comment by mystor** [comment]
The example with the iterator works OK while the one with the slice does not due to a slight difference in the function signatures. `slice::is_sorted_by_key` takes `for<'a> FnMut(&'a T) -> K`, meaning that `K` cannot borrow the argument, whereas `Iterator::is_sorted_by_key` takes `FnMut(Self::Item) -> K`, meaning that `K` can borrow the argument. This was likely done for symmetry with `slice::sort_by_key` (which needs to use a `for<'a>` closure as elements will be re-ordered), but isn't necessary for `is_sorted_by_key` which could be relaxed to take the same lifetime as `self`, due to there being no mutations. This would also apply to `slice::is_sorted_by`.

The rough change would probably look like:

```rust
pub fn is_sorted_by<'a, F>(&'a self, mut compare: F) -> bool
where
    F: FnMut(&'a T, &'a T) -> Option<Ordering>,
{
    self.iter().is_sorted_by(|a, b| compare(*a, *b))
}

pub fn is_sorted_by_key<'a, F, K>(&'a self, f: F) -> bool
where
    F: FnMut(&'a T) -> K,
    K: PartialOrd,
{
    self.iter().is_sorted_by_key(f)
}
```

**Comment by leonardo-m** [comment]
If slice::sort uses F: FnMut(&T, &T) -> Ordering, then perhaps is_sorted_by could use the same.

**Comment by nm17** [comment]
A suggestion: Rust should raise a compiler error when is_sorted or one of the related functions is called on keys or values of a `HashMap`, since depending on their order is UB.

**Comment by clarfonthey** [comment]
Depending on the order of items in a HashMap isn't UB; the order is just not guaranteed. The difference here is one thing is unsafe and one is just a bad idea.

**Comment by rdrpenguin04** [comment]
"Revisit test suite" can probably be checked off due to #112699.

I feel like this could probably be stabilized as-is, and more could be added in later.

**Comment by WaffleLapkin** [comment]
**Nominating for `T-libs-api`**: this API existed for a while, and the only question left seems to be whatever `cmp::is_sorted` would be a better API option. Thus, I'm proposing stabilization :)

Note: while trying to check the implementation history I've updated the [issue text](https://github.com/rust-lang/rust/issues/53485#issue-351885772) with the newer format, hopefully it makes it easier to understand what this issue tracks, exactly.

While doing that I've also found two questions in the comments that did not seem to be fully answered:
- https://github.com/rust-lang/rust/issues/53485#issuecomment-509026850 — questions whatever `Option<Ordering>` is the right thing to ask for `is_sorted_by`
- https://github.com/rust-lang/rust/issues/53485#issuecomment-633001259 — questions whatever we should have a function that returns a position up to which everything is sorted

**Comment by Jules-Bertholet** [comment]
This was already mentioned earlier, but `is_sorted_by_key`, like all the `by_key` methods, suffers from #34162; a fix would require removing the `K` type parameter.

**Comment by Amanieu** [comment]
Now that the concerned from the [previous](https://github.com/rust-lang/rust/issues/53485#issuecomment-1726616457) stabilization report have been address, let's stabilize this.

This specifically proposes to keep the current API and not add `std::cmp::is_sorted` which was proposed as an alternative in the RFC.

It also doesn't do anything to address #34162, but that already applies to the existing `_by_key` functions such as `sort_by_key`. Any solution for that issue could be applied to all `_by_key` functions.

@rfcbot merge

```rust
impl<T> [T] {
    pub fn is_sorted(&self) -> bool
    where
        T: PartialOrd;

    pub fn is_sorted_by<'a, F>(&'a self, mut compare: F) -> bool
    where
        F: FnMut(&'a T, &'a T) -> bool;

    pub fn is_sorted_by_key<'a, F, K>(&'a self, f: F) -> bool
    where
        F: FnMut(&'a T) -> K,
        K: PartialOrd;
}
```

```rust
// core::iter

pub trait Iterator {
    fn is_sorted(self) -> bool
    where
        Self: Sized,
        Self::Item: PartialOrd;

    fn is_sorted_by<F>(mut self, compare: F) -> bool
    where
        Self: Sized,
        F: FnMut(&Self::Item, &Self::Item) -> bool;

    fn is_sorted_by_key<F, K>(self, f: F) -> bool
    where
        Self: Sized,
        F: FnMut(Self::Item) -> K,
        K: PartialOrd;
}
```

**Comment by BurntSushi** [comment]
My first reaction to these APIs was to find it odd that `Ordering` wasn't used. I immediately wondered why it was inconsistent with the existing `sort_by` APIs. So I read up and found this:

> * Using `bool` rather than `Option<Ordering>` for the return value of the closure to `is_sorted_by` seems clearer and more flexible. E.g. then you can use `a < b` or `a <= b` depending on whether it should be strictly increasing or not, etc.

The `Option<Ordering>` made me realize that the `is_sorted_by` requires `PartialOrd` and not `Ord`. Indeed, if we want to use `PartialOrd` here, then returning an `Option<Ordering>` is pretty annoying. But this bugs me a little bit, because AFAIK, we use `Ordering` everywhere in the standard library whenever we do anything oriented around comparisons. Is this the first API where we shed that and use `bool` instead? And I don't think there's anything in the name of these functions that points callers toward what the `bool` is supposed to mean. Is it less than? Or greater than? ¯\_(ツ)_/¯

If we used `Ord` instead of `PartialOrd`, then my feeling is that returning `Ordering` is ergonomic enough to go with it instead of a `bool` here. But I assume the main practical benefit of using `PartialOrd` is that this will work on floating point types. But I'm not sure that's the right trade-off given the existence of [`f64::total_cmp`](https://doc.rust-lang.org/std/primitive.f64.html#method.total_cmp).

What pieces am I missing here? Is there something more definitive pushing us toward `PartialOrd` here?

**Comment by BurntSushi** [comment]
@rfcbot concern inconsistent-with-sort

I'm registering this concern because I'm not a fan of the inconsistency between these APIs and our existing `sort` APIs. I'm not saying they _have_ to be consistent, but I don't think I'm seeing a compelling justification for doing something different here.

> If you have `PartialOrd` types then you should probably be using `is_sorted` or `is_sorted_by_key`. `is_sorted_by` is intended for cases where you want full control to answer the question "are there 2 elements in order?". It makes sense to return a `bool` here.

Yeah but as #34162 demonstrates, you often need to use `is_sorted_by` because `is_sorted_by_key` doesn't work due to a limitation in its type signature and/or Rust's expressiveness (depending on how you view it I suppose). This happens a lot in practice for me where the natural thing to use is indeed `sort_by_key`, but I can't because of the lifetime issue. So I end up needing to use `sort_by`. I assume the same will be true for `is_sorted_by` and `is_sorted_by_key`.

> I read it as "is it sorted how you want", asc, desc, in a circle, etc.

I guess this makes sense. It wasn't what I saw on first impression, but I suppose that can be fixed with docs.

> Leaving the meaning blank actually feels more flexible; you get to use .partial_cmp or .cmp in your closure and give it readable semantics rather than the tricky-to-skim-read Ordering return value (IMO just as hard as -1,0,+1 returns in the JS' optional Array.sort argument). E.g., returning Ordering::Greater for lower numbers when trying to sort in a descending fashion is weird.

I don't think I want to get into a "what's easier to read as a comparator, `bool` or `Ordering`" discussion here. Rust's standard library uses the latter pretty much everywhere as far as I can tell whenever comparisons come up. Why are we doing it differently here? Are we going to commit to the abdication of `Ordering` in favor of `bool` for any future comparator oriented APIs?

I think I overall need this inconsistency spelled out a bit more for me. Like I said, I'm not trying to say we should be consistent for consistency's sake. ("A foolish consistency is the hobgoblin of little minds.") But that consistency has _value_ and I'm not sure we're gaining enough to justify throwing away that value here.

**Comment by joshtriplett** [comment]
I do personally think that the function returning `bool` is easier to write (you can write a closure that just uses `<=` or `<`, rather than having to translate to an Ordering), but I think the consistency is important as well.

Suggestion: could we go ahead and do an FCP to stabilize `is_sorted` and `is_sorted_by_key`, which aren't blocked by continuing to debate either the name or signature of `is_sorted_by`?

**Comment by joshtriplett** [comment]
Nominating for us to sort this out.

**Comment by BurntSushi** [comment]
> Suggestion: could we go ahead and do an FCP to stabilize `is_sorted` and `is_sorted_by_key`, which aren't blocked by continuing to debate either the name or signature of `is_sorted_by`?

This could lead to a frustrating state of affairs ("I can almost use `is_sorted_by_key` but actually need `is_sorted_by` which isn't stable") but I'm fine with it.

**Comment by m-ou-se** [comment]
> I don't think I want to get into a "what's easier to read as a comparator, `bool` or `Ordering`" discussion here. Rust's standard library uses the latter pretty much everywhere as far as I can tell whenever comparisons come up. Why are we doing it differently here? Are we going to commit to the abdication of `Ordering` in favor of `bool` for any future comparator oriented APIs?

Responding to the "why are we doing it differently here?" part:

We use `Ordering` or `Option<Ordering>` when all three/four possibilities (less, equal, greater, unordered) have some meaning. In this case, we only check whether two elements are "sorted" yes or no, which is a boolean decision. If we used a `Ordering` API here, then switching between `<` and `<=` gets really awkward, having to swap `Less` for `Equal` or the other way around..

**Comment by m-ou-se** [comment]
Perhaps we should instead call it `.pairwise().all(|a, b| a > b)` rather than `.is_sorted_by(|a, b| a > b)`, to avoid the inconsistency..

**Comment by Amanieu** [comment]
Additionally, there is precedent for comparison closures returning a `bool`, such as `Vec::dedup_by` and `[T]::partition_point`. We discussed this again at the @rust-lang/libs-api meeting and everyone there agreed that returning `bool` in this case is better overall than returning an `Ordering` for the reasons previously listed.

**Comment by SimonSapin** [comment]
We already have `slice.array_windows().all(|[a, b]| a <= b)` but that’s even less discoverable

**Comment by BurntSushi** [comment]
> In this case, we only check whether two elements are "sorted" yes or no, which is a boolean decision.

I found this phrasing compelling. Thank you!

@rfcbot resolve inconsistent-with-sort

**Comment by ahakkar** [comment]
As a late note, I find it greatly infuriating that sorted_by requires a `F: FnMut(&T, &T) -> Ordering`, while is_sorted_by requires a F: `FnMut(&'a T, &'a T) -> bool`

It makes no sense at all for the "same" functions to require completely different input. Similar functions should definitely always require similar input for consistency.
